fdb-extensions/src/main/java/com/apple/foundationdb/async/RankedSet.java

```diff
@@ -220,32 +391,110 @@ public RankedSet(Subspace subspace, Executor executor) {
-            return CompletableFuture.allOf(prevCount, newPrevCount)
-                    .thenApply(vignore2 -> {
-                        long count = prevCount.join() - newPrevCount.join() + 1;
-                        tr.set(subspace.pack(Tuple.from(level, prevKey)), encodeLong(newPrevCount.join()));
-                        tr.set(subspace.pack(Tuple.from(level, key)), encodeLong(count));
-                        return null;
-                    });
+            return prevCount.thenAcceptBoth(newPrevCount, (prev, newPrev) -> {
+                long count = prev - newPrev + 1;
+                tr.set(subspace.pack(Tuple.from(level, prevKey)), encodeLong(newPrev));
+                tr.set(subspace.pack(Tuple.from(level, key)), encodeLong(count));
+            });
         });
     }
 
+    /**
+     * Removes a key from the set.
+     * @param tc the transaction to use to access the database
+     * @param key the key to remove
+     * @return a future that completes to {@code true} if the set was modified, that is, if the key was present before this operation
+     */
+    public CompletableFuture<Boolean> remove(TransactionContext tc, byte[] key) {
+        checkKey(key);
+        return tc.runAsync(tr ->
+                countCheckedKey(tr, key)
+                        .thenCompose(count -> {
+                            if (count == null || count <= 0) {
+                                return READY_FALSE;
+                            }
+                            // This works even if the current set does not track duplicates but duplicates were added
+                            // earlier by one that did.
+                            final boolean duplicate = count > 1;
+                            final int nlevels = config.getNLevels();
+                            final List<CompletableFuture<Void>> futures = new ArrayList<>(nlevels);
+                            for (int li = 0; li < nlevels; ++li) {
+                                final int level = li;
+
+                                final CompletableFuture<Void> future;
+
+                                if (duplicate) {
+                                    // Always subtract one, never clearing a level count key.
+                                    // Concurrent requests both subtracting one when the count is two will conflict
+                                    // on the level zero key, which was read to detect that this is a duplicate.
+                                    // So it should not be possible for a count to go to zero.
+                                    Function<byte[], Void> decrement = k -> {
+                                        tr.mutate(MutationType.ADD, subspace.pack(Tuple.from(level, k)), encodeLong(-1));
+                                        return null;
+                                    };
+                                    if (level == 0) {
+                                        decrement.apply(key);
```

Comment:
I suppose rather than decrementing, this could explicitly set the key to `encodeLong(count - 1)`. I think the only thing this would do is underscore that we *aren't* actually making use of the ability of atomic operations to allow for concurrent updates to the level 0 keys.

Reply:
Yes, that does seem more obvious and in line with the preceding comment.